### ISSUE TITLE
usFrameWorksExamplesDriver hangs when quitting after stopping system bundle

### ISSUE THREAD

**Issue by ksubramz** [CppMicroServices/CppMicroServices#156]
### Steps to repro

```
λ usFrameworkExamplesDriver.exe
> u 0
> q
```

It hangs here and I have to press [Ctrl-C] to exit out of the driver program. (Tested on Win 10)

Under any other scenario "q" exits gracefully and returns us back to the shell.

**Comment by jeffdiclemente** [comment]
uninstalling the system bundle should throw an exception. Its not allowed, per the OSGi spec.

**Comment by ksubramz** [comment]
@jeffdiclemente  suggested to try the following snippet just before exiting the driver program to see if this is a similar bug he is working on
```
  framework.Stop();
  framework.WaitForStop(std::chrono::seconds::zero())
```
However, it still hangs in WaitForStop() method.

**Comment by ksubramz** [comment]
As @jeffdiclemente said, uninstalling the system bundle should throw an exception acc. to the OSGi spec. 
 
My tentative idea of a fix is to 
1. implement Framework::Uninstall() and throw an exception if this is called per spec
2. Check if GetBundleId() == 0 in Bundle::Uninstall() and throw the same exception
(Of course, I maybe wrong about the fix :-) )

However, even if this is implemented, the reason behind the aforementioned hang is still unknown.

**Comment by ksubramz** [comment]
Implementing the exception in Bundle::Uninstall() locally led me to this finding that we don't "uninstall" the bundle when we do "u 0" in the driver program. According to the help 

```
> h
h               This help text
l <id | name>   Start the bundle with id <id> or name <name>
u <id>          Stop the bundle with id <id>
s               Print status info
```

So we are simply stopping it. I erroneously associated the letter 'u' with 'uninstall'. 

I have changed the title to reflect this.

**Comment by ksubramz** [comment]
Now, I check if the bundle id is 0 and if so, I do the WaitForStop() like Jeff suggested i.e.

```
        try
        {
          bundle.Stop();
		  if (id == 0)
		  {
		    framework.WaitForStop(std::chrono::seconds::zero());
		  }
        }
```
And this doesn't hang! 
```
λ usFrameworkExamplesDriver.exe
> u 0
> q
```
(returns to shell gracefully)

**Comment by ksubramz** [comment]
It turns out a framework.WaitForStop() immediately after bundle.Stop() (refer to the previous update) works as desired but a framework.WaitForStop() down below (just before the return), hangs. Jeff surmised it maybe due to a race condition.

**Comment by saschazelzer** [comment]
I believe the issues you found have been fixed in #130 . For the hangs, I couldn't reproduce them. But I am not sure I fully understand what combinations you tries. If you added a `WaitForStop()` call before returning from main but without actually calling `framework.Stop()`, then this would hang for sure.

Feel free to re-open this issue if you think there is still something to fix here.

**Comment by ksubramz** [comment]
Just to close the loop, you are right @saschazelzer. The issue seems to have been fixed with your changes and I no longer encounter the hang.
